internal/candidate_runner/runner.go

```diff
@@ -0,0 +1,346 @@
+	"github.com/planetlabs/draino/internal/scheduler"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/planetlabs/draino/internal/groups"
+	"github.com/planetlabs/draino/internal/kubernetes"
+	"github.com/planetlabs/draino/internal/kubernetes/drain"
+	"github.com/planetlabs/draino/internal/kubernetes/index"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Make sure that the drain runner is implementing the group runner interface
+var _ groups.Runner = &candidateRunner{}
+
+var now = time.Now
+
+type NodeSorters []scheduler.LessFunc[*corev1.Node]
+type NodeIteratorFactory func([]*corev1.Node, NodeSorters) scheduler.ItemProvider[*corev1.Node]
+
+// candidateRunner implements the groups.Runner interface and will be used to drain nodes of the given group configuration
+type candidateRunner struct {
+	client               client.Client
+	logger               logr.Logger
+	clock                clock.Clock
+	retryWall            drain.RetryWall
+	sharedIndexInformer  index.GetSharedIndexInformer
+	runEvery             time.Duration
+	cordonFilter         kubernetes.PodFilterFunc
+	eventRecorder        kubernetes.EventRecorder
+	objectsStore         kubernetes.RuntimeObjectStore
+	nodeLabelsFilterFunc kubernetes.NodeLabelFilterFunc
+	globalConfig         kubernetes.GlobalConfig
+
+	maxSimultaneousCandidates int
+	dryRun                    bool
+
+	nodeSorters         NodeSorters
+	nodeIteratorFactory NodeIteratorFactory
+	drainSimulator      drain.DrainSimulator
+}
+
+func (runner *candidateRunner) Run(info *groups.RunnerInfo) error {
+	ctx, cancel := context.WithCancel(info.Context)
+
+	// TODO if we add metrics addociated with that key, when the group is closed we should purge all the series associated with that key (cleanup-gauges with groupKey=...)?
+	runner.logger = runner.logger.WithValues("groupKey", info.Key)
+	// run an endless loop until there are no drain candidates left
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
+		nodes, err := index.GetFromIndex[corev1.Node](ctx, runner.sharedIndexInformer, groups.SchedulingGroupIdx, string(info.Key))
+		// in case of an error we'll just try it again
+		if err != nil {
+			runner.logger.Error(err, "cannot get nodes for group", "group_key", info.Key)
+			return
+		}
+
+		// TODO add metric to track amount of nodes in the group
+		if len(nodes) == 0 {
+			// If there are no candidates left, we'll stop the loop
+			runner.logger.Info("no nodes in group left, stopping.", "group_key", info.Key)
+			cancel()
+			return
+		}
+
+		// TODO add metrics after each filter ?
+		// TODO is there a better order for filter that would use less time/resources ?
+
+		// filter nodes that are already candidate
+		var maxReached bool
+		var alreadyCandidateNodes []*corev1.Node
+		nodes, alreadyCandidateNodes, maxReached = runner.checkAlreadyCandidates(ctx, nodes)
```

Comment:
Can't you just use `:=`?
```suggestion
		nodes, alreadyCandidateNodes, maxReached := runner.checkAlreadyCandidates(ctx, nodes)
```

Reply:
We can we have a partial definition in that case: nodes is already defined. But that is ok.